iceoryx2-ffi/cxx/include/iox2/sample_mut.hpp

```diff
@@ -186,27 +194,55 @@ inline auto SampleMut<S, Payload, UserHeader>::user_header_mut() -> T& {
 }
 
 template <ServiceType S, typename Payload, typename UserHeader>
+template <typename T, typename>
 inline auto SampleMut<S, Payload, UserHeader>::payload() const -> const Payload& {
     const void* ptr = nullptr;
-    size_t payload_len = 0;
+    size_t number_of_bytes = 0;
 
-    iox2_sample_mut_payload(&m_handle, &ptr, &payload_len);
-    IOX_ASSERT(sizeof(Payload) <= payload_len, "");
+    iox2_sample_mut_payload(&m_handle, &ptr, &number_of_bytes);
+    IOX_ASSERT(sizeof(Payload) <= number_of_bytes, "");
 
     return *static_cast<const Payload*>(ptr);
 }
 
 template <ServiceType S, typename Payload, typename UserHeader>
+template <typename T, typename>
 inline auto SampleMut<S, Payload, UserHeader>::payload_mut() -> Payload& {
     void* ptr = nullptr;
-    size_t payload_len = 0;
+    size_t number_of_bytes = 0;
 
-    iox2_sample_mut_payload_mut(&m_handle, &ptr, &payload_len);
-    IOX_ASSERT(sizeof(Payload) <= payload_len, "");
+    iox2_sample_mut_payload_mut(&m_handle, &ptr, &number_of_bytes);
+    IOX_ASSERT(sizeof(Payload) <= number_of_bytes, "");
 
     return *static_cast<Payload*>(ptr);
 }
 
+template <ServiceType S, typename Payload, typename UserHeader>
+template <typename T, typename>
+inline auto SampleMut<S, Payload, UserHeader>::payload_slice() const -> const Payload {
+    const void* ptr = nullptr;
+    size_t number_of_bytes = 0;
+
+    iox2_sample_mut_payload(&m_handle, &ptr, &number_of_bytes);
+
+    size_t number_of_elements = number_of_bytes / sizeof(typename Payload::ValueType);
```

Comment:
I realized we need to carefully test this ... test it very carefully.

The thing is, that the `message_type_details` are used to calculate the size and alignment of the raw memory in `Publisher::loan_uninit`, see `publisher.rs:835`. With this in mind, it is actually the always the number of elements and we never need to divide as an extra step!

So the C API is with the message type details also strongly typed. When this has been set, the C API slice loan calls behave exactly like the strongly typed rust pendants. The actual Rust implementation never uses `sizeof(T)` or `alignof(T)` to calculate anything - it always uses the message type details and those message type details are constructed by calling `sizeof(T)` and `alignof(T)` but in the C++ pub sub builder you override this correctly with the C API and thus turn the C API into a strongly typed API. So the C loan_slice call actually returns you the number of elements and never the number of bytes.

Sorry for the huge confusion from my side here!

So could you adjust please all code parts with this line, where you divide some number of bytes with `sizeof(typename Payload::ValueType)`.

Reply:
@elfenpiff It's not exactly clear what to change. The tests are passing with this implementation, which means `loan_slice` is returning the number of bytes in the `[u8]`. 

The code is determining this using the `.len()` method of the slice, which is the number of elements in the u8 slice: https://github.com/eclipse-iceoryx/iceoryx2/pull/492/files#diff-de3b4aa52551d38602a11d9d6689468c205b0acfd84b3702c5aed8a7442c212cR237

This value would need to be divided by the size of the type in Rust if we were to remove it from the C++ binding. Is this information available at the line linked?